packages/i18n/src/create-i18n.js

```diff
@@ -1,3 +1,8 @@
+/**
+ * WordPress dependencies
+ */
+import { applyFilters } from '@wordpress/hooks';
```

Comment:
The `createI18n` function shouldn't always use the global `applyFilters` function. It should allow for supplying a custom instance of the `Hooks` object returned by `createHooks`.

The `hooks` and `i18n` packages have a similar structure:
- there is a `createI18n` or `createHooks` function that creates a `I18n` or `Hooks` object that has methods like `__` or `applyFilters`
- the program that uses these libraries can create multiple instances of these objects that are fully independent from each other and don't share any data. Imagine a Node.js server that returns server-generated pages, serving multiple users, and handling the requests in parallel. Each request has its own, isolated context.
- both libraries create a "default" instance of their respective objects, and exports their methods as functions. That's what `__` and `applyFilters` are -- methods of one specific object instance that are used as functions.

To keep the separation between default singletons and custom instances, `createI18n` should take a new `hooks` parameter:
```js
createI18n( initialData, initialDomain, hooks )
```
use this `hooks` parameter to get to the right `applyFilters`:
```js
return String( hooks.applyFilters( ... ) );
```
and the `default-i18n.js` module should configure it with the default hooks:
```js
import { applyFilters } from '@wordpress/hooks';

// if the hooks package exported the `Hooks` object itself, we would pass it as is and
// wouldn't have to create a 'fake' one with `{ applyFilters }`
const i18n = createI18n( null, null, { applyFilters } );

Reply:
@jsnajdr I've updated the PR to follow this approach. The `hooks` arg is optional, and no filters will be called if it is not provided. default-i18n provides a copy of `applyFilters` from `@wordpress/hooks` when creating the default instance. Feedback welcome.